flocker/testtools/_flaky.py

```diff
@@ -22,10 +43,274 @@ def flaky(jira_key):
 
     # TODO (see FLOC-3281):
     # - allow specifying which exceptions are expected
-    # - retry on expected exceptions
-    #   - parametrize on retry options
     # - provide interesting & parseable logs for flaky tests
 
+    if isinstance(jira_keys, unicode):
+        jira_keys = [jira_keys]
+
+    annotation = _FlakyAnnotation(
+        jira_keys=pset(jira_keys), max_runs=max_runs, min_passes=min_passes)
+
     def wrapper(test_method):
+        existing_flaky = getattr(test_method, _FLAKY_ATTRIBUTE, None)
+        if existing_flaky is None:
+            note = annotation
+        else:
+            note = _combine_flaky_annotation(annotation, existing_flaky)
+        setattr(test_method, _FLAKY_ATTRIBUTE, note)
         return test_method
+
     return wrapper
+
+
+def _get_flaky_annotation(case):
+    """
+    Get the flaky decoration detail from ``case``.
+
+    :param TestCase case: A test case that might have been decorated with
+        @flaky.
+    :return: ``None`` if not flaky, or a ``pmap`` of the flaky test details.
+    """
+    # XXX: Alas, there's no public way of doing this:
+    # https://bugs.launchpad.net/testtools/+bug/1517867
+    method = case._get_test_method()
+    return getattr(method, _FLAKY_ATTRIBUTE, None)
+
+
+class _FlakyAnnotation(PClass):
+
+    max_runs = field(int, mandatory=True,
+                     invariant=lambda x: (x > 0, "must run at least once"))
+    min_passes = field(int, mandatory=True,
+                       invariant=lambda x: (x > 0, "must pass at least once"))
+    jira_keys = pset_field(unicode, optional=False)
+
+    __invariant__ = lambda x: (
+        (x.max_runs >= x.min_passes, "Can't pass more than we run"),
+        (len(x.jira_keys) > 0, "Must provide a jira key"),
+    )
+
+    def to_dict(self):
+        return {
+            'max_runs': self.max_runs,
+            'min_passes': self.min_passes,
+            'jira_keys': set(self.jira_keys),
+        }
+
+
+def _combine_flaky_annotation(flaky1, flaky2):
+    """
+    Combine two flaky annotations.
+    """
+    return _FlakyAnnotation(
+        jira_keys=flaky1.jira_keys | flaky2.jira_keys,
+        max_runs=max(flaky1.max_runs, flaky2.max_runs),
+        min_passes=max(flaky1.min_passes, flaky2.min_passes),
```

Comment:
:grey_question: So, I think this PR has pretty much blown it's complexity budget so don't necessarily take action on this point, but I have some interesting observations here.

```
@flaky('XXX', min_passes=1, max_runs=100) # If this ever passes it means the code is probably fine.
@flaky('YYY', min_passes=3, max_runs=3) # This test is quick, and I think all the flakes are fixed, run it a bit more frequently for the next week to verify.
def test_...
```
Resolves kinda funny, but shame on us if we ever set up a situation like that.

More interestingly:
```
@flaky('XXX')
@flaky('YYY', min_passes=1, max_runs=2) # test is almost stable, and failures take 5 min to time out.
```
resolves oddly as well.

My hunch is using the values from whichever one has the higher standards is the right call:
```
def ratio(f):
   return float(f.min_passes)/float(f.max_runs)

higher_standards = flaky1 if ratio(flaky1) > ratio(flaky2) else flaky2
```
but... as I said complexity budget running low and this is just test infrastructure, so what you have here certainly is good enough.